renormalizer/mps/svd_qn.py

```diff
@@ -1,10 +1,67 @@
 # -*- coding: utf-8 -*-
 # Author: Jiajun Ren <jiajunren0522@gmail.com>
+import logging
 
 import scipy.linalg
 
 from renormalizer.mps.backend import np
 
+logger = logging.getLogger(__name__)
+
+
+def optimized_svd(a, full_matrices):
+    # optimize performance when ``full_matrices=True`` and the shape of ``a`` is extremely unbalanced
+    # The idea is to construct only a limited number of orthogonal basis rather than all of them
+    # (which are not necessary in most cases)
+    m, n = a.shape
+
+    # whether do the optimization
+    # here 1/3 and 3 are only empirical
+    opt = not (1 / 3 < m / n < 3)
+
+    # if opt, always set ``full_matrices=False``
+    try:
+        U, S, Vt = scipy.linalg.svd(
+            a,
+            full_matrices=full_matrices and not opt,
+            lapack_driver="gesdd",
+        )
+    except scipy.linalg.LinAlgError:
+        logger.warning("Csvd failed to converge")
+        U, S, Vt = scipy.linalg.svd(
+            a,
+            full_matrices=full_matrices and not opt,
+            lapack_driver="gesvd",
+        )
+    if not opt:
+        return U, S, Vt
+
+    # if opt, add n  additional basis assuming  2 * n < m
+    if m < n:
+        Vt = add_orthonormal_basis(Vt.T).T
+    elif n < m:
+        U = add_orthonormal_basis(U)
+    else:
+        assert False
+    return U, S, Vt
+
+
+def add_orthonormal_basis(u):
+    # add `n` basis. `n` is empirical
+    m, n = u.shape
+    assert 2 * n < m
+    assert np.allclose(u.T.conj() @ u, np.eye(n))
+    res = np.zeros((m, 2 * n), dtype=u.dtype)
+    res[:, :n] = u
+    for i in range(n):
+        a = np.random.rand(m)
+        a = a - res @ (res.T.conj() @ a)
+        a /= np.linalg.norm(a)
+        res[:, n + i] = a
```

Comment:
For higher efficiency with vectorization, I think about another algorithm to construct `res` without the manual Gram-Schmidt orthogonalization.
```
a = np.random.rand(m,n)
a = a - res @ (res.T.conj() @ a)
q, r = scipy.linalg.qr(a, mode='economic') 
res[:, n:] =  q
```

Reply:
Do you mean:
```
a = np.random.rand(m,n)
a = a - u @ (u.T.conj() @ a)
q, r = scipy.linalg.qr(a, mode='economic') 
res[:, n:] =  q
```
I think there should be a measurable speed up, I'll carry out some tests for sure.

Reply:
For large matrices (100000 * 100), old routine costs 1.64 s, while the QR version costs 0.682 s.
For small matrices (100 * 10), old routine costs 420 mus, while the QR version costs 301 mus.

Reply:
That's good.